### DISCUSSION TITLE
Ways to speed up Raspberry Pi boot up time

### DISCUSSION THREAD

**Discussion by intelliant01** [ltsp/ltsp#577] [Q&A] [answered]
Am using RPi 4B 2/4GB as clients with a Ubuntu based LTSP server and Pi image based on guidance given herein - https://github.com/ltsp/ltsp/discussions/546

The current boot time for Pi clients is around 85-90 seconds while using a gigabit switch.

Need suggestions on how to reduce the boot time.

**Comment by enaut** [comment, marked as answer]
Disable services not needed, try to get the main offenders with `systemd-analyze plot > graph.svg`. It is basically the same as improving the non ltsp bootup time. The boot from network just adds another 15 seconds or so.

**Reply by intelliant01** [reply]
Thanks @enaut. I shall try this. Have found a few services that I could mask. But that shall not be enough to bring it under a minute.

For x86 clients I served the kernel and initrd via http which helped get the boot time under 40 seconds. But couldn't get a way to do so for the Pi. Any ideas?

**Reply by intelliant01** [reply]
Have not been able to improve beyond 85-87 seconds.
Herein are attached the outputs of `systemd-analyze` and `systemd-analyze blame`. 
Do you spot any service(s), masking which would help?

[blame1.txt](https://github.com/ltsp/ltsp/files/7361733/blame1.txt)
[time_taken1.txt](https://github.com/ltsp/ltsp/files/7361734/time_taken1.txt)

**Reply by enaut** [reply]
I'd say if possible remove snap... But other than that I guess it's just that it is that ubuntu is probably  not the fastest option for the rpi

**Comment by alkisg** [comment]
Raspberry Pi OS images are more optimized for Raspberries than the Ubuntu ones, as they take into account their slow CPUs and SD cards etc. So they boot faster over the network as well. For example, they do not even need an initrd, which cuts e.g. 10 seconds from the boot time.

With LTSP5, I tested and added support for Ubuntu MATE on Raspberries. But with the new LTSP I have only tested with the official Raspberry Pi OS images, nothing else!

Although the boot time isn't as important as later on, if they can launch firefox or libreoffice from the network as fast (or faster) as from an SD card.

**Reply by intelliant01** [reply]
> Although the boot time isn't as important 

Absolutely agreed, but there are system admins and teachers at schools who tend to hang on to this stat. Just want to give it my best shot.

> as later on, if they can launch firefox or libreoffice from the network as fast (or faster) as from an SD card.

Yes, the speeds post boot are good enough.

**Reply by intelliant01** [reply]
> they do not even need an initrd, which cuts e.g. 10 seconds from the boot time.

Indeed that and vmlinuz seem a big time hog.

**Comment by alkisg** [comment]
Btw where are you counting from, the initial rainbow screen and up to the user desktop?
Or e.g. from "loading vmlinuz" to lightdm?
And how long does it take booting from an SD card?
And which rpi version, 2, 3, 4?
And which desktop environment?

**Reply by intelliant01** [reply]
Ubuntu Mate 20.04
RPi 4B 4G
Gigabit network

Measuring from power on to lightdm login screen appearance.

SD card with lot more services than the ltsp Pi image takes 66 seconds.

**Comment by alkisg** [comment]
When vmlinuz/initrd are downloaded, the networking is 100Mbps, not gigabit. And TFTP as a protocol is very slow, try switching to HTTP. Or if these seconds really matter to you, you can put just vmlinuz/initrd in local SD cards.

If you don't count the vmlinuz/initrd downloading time, is the rest of the boot speed the same?

**Reply by intelliant01** [reply]
Couldn't figure out how to switch to http for the Pi. Had done it for x86 and results were great. Any pointers that I can explore?

Yes, initrd and vmlinuz take around ~20 second. Thus cutting that down would make the boot time same as that of sd card.

**Comment by alkisg** [comment]
Well there's https://github.com/ipxe/pipxe but that would add its own time to the boot sequence, not sure if the final result would be less or more.

My rpi2 + bullseye netboots in 50 seconds, I guess the rpi4 would be a lot faster.

You could check if the Ubuntu kernel has networking built in, in which case you could completely omit loading the initrd, like Raspberry Pi OS does.

**Reply by intelliant01** [reply]
Thanks for the ideas.

> Well there's https://github.com/ipxe/pipxe ....

Will bring an SD card into the equation that I want to avoid at all costs.

> You could check if the Ubuntu kernel has networking built in, in which case you could completely omit loading the initrd, like Raspberry Pi OS does.

Omitting loading the initrd, results in a kernel panic. 

May have to explore this further.
